packages/flutter/lib/src/widgets/app_bar.dart

```diff
@@ -0,0 +1,238 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+
+import 'framework.dart';
+
+abstract class SliverAppBarDelegate {
+  /// Abstract const constructor. This constructor enables subclasses to provide
+  /// const constructors so that they can be used in const expressions.
+  const SliverAppBarDelegate();
+
+  Widget build(BuildContext context, double shrinkOffset);
+
+  double get maxExtent;
+
+  bool shouldRebuild(@checked SliverAppBarDelegate oldDelegate);
+}
+
+class SliverAppBar extends StatelessWidget {
+  SliverAppBar({
+    Key key,
+    @required this.delegate,
+    this.pinned: false,
+    this.floating: false,
+  }) : super(key: key) {
+    assert(delegate != null);
+    assert(pinned != null);
+    assert(floating != null);
+    assert(!pinned || !floating);
+  }
+
+  final SliverAppBarDelegate delegate;
+
+  final bool pinned;
+
+  final bool floating;
+
+  @override
+  Widget build(BuildContext context) {
+    if (pinned)
+      return new _SliverPinnedAppBar(delegate: delegate);
+    if (floating)
+      return new _SliverFloatingAppBar(delegate: delegate);
+    return new _SliverScrollingAppBar(delegate: delegate);
+  }
+
+  @override
+  void debugFillDescription(List<String> description) {
+    super.debugFillDescription(description);
+    description.add('delegate: $delegate');
+    List<String> flags = <String>[];
+    if (pinned)
+      flags.add('pinned');
+    if (floating)
+      flags.add('floating');
+    if (flags.isEmpty)
+      flags.add('normal');
+    description.add('mode: ${flags.join(", ")}');
+  }
+}
+
+class _SliverAppBarElement extends RenderObjectElement {
```

Comment:
This class looks pretty generic.  We should watch to see if this pattern recurs.